File: dart/dart-sample/sample/oop/inheritance.dart
void main(List<String> args) {
  var dog = Dog('snoopy');
  dog.speak();
}

abstract class Animal {
  // abstract methods
  void speak();
}

class Dog extends Animal {
  String name;
  Dog(this.name);

  @override
  void speak() {
    print('Dog $name sparks');
  }
}

File: nodejs/typescript-sample/oop/inheritance.ts
abstract class Animal {
  protected name: string
  constructor(name: string) {
    this.name = name
  }

  move (): void {
    console.log(`${this.name} moves`)
  }
  abstract speak (): void
}

class Cat extends Animal {
  move (): void {
    console.log(`${this.name} moves with four legs`)
  }
  speak (): void {
    console.log(`${this.name} meows`)
  }
}

const cat = new Cat('kitty')
cat.move()
cat.speak()
